src/utils/helpers.test.ts

```diff
@@ -157,3 +158,88 @@ describe('formatConversationTitle', () => {
     ).toBe('This is a ....');
   });
 });
+
+describe('groupAndFilterConversations', () => {
+  let mockSearchHistories: SearchableChatHistories;
+  const now = new Date('2024-02-13T12:00:00Z').getTime();
+
+  beforeEach(() => {
+    vi.spyOn(Date.prototype, 'getTime').mockImplementation(() => now);
+
+    mockSearchHistories = {
+      1: {
+        title: 'Bitcoin Discussion',
+        lastSaved: now - 1000 * 60 * 60 * 2,
+        messages: [{ role: 'user', content: 'Can I have bitcoin advice?' }],
+      },
+      2: {
+        title: 'Ethereum Chat',
+        lastSaved: now - 1000 * 60 * 60 * 25,
+        messages: [{ role: 'user', content: 'Where can I buy ETH?' }],
+      },
+      3: {
+        title: 'Blockchain Overview',
+        lastSaved: now - 1000 * 60 * 60 * 24 * 5,
+        messages: [
+          { role: 'user', content: 'What is the history of blockchain?' },
+        ],
+      },
+      4: {
+        title: 'Party planning tips',
+        lastSaved: now - 1000 * 60 * 60 * 24 * 6,
+        messages: [{ role: 'user', content: 'I need help planning.' }],
+      },
+    };
+  });
+
+  afterEach(() => {
+    vi.restoreAllMocks();
+  });
+
+  it('should filter conversations based on search term', () => {
+    const filtered = groupAndFilterConversations(
+      mockSearchHistories,
+      'bitcoin',
+    );
+    expect(Object.keys(filtered)).toEqual(['Today']);
+    expect(filtered['Today'][0].title).toBe('Bitcoin Discussion');
+  });
+
+  it('should be case insensitive when filtering', () => {
+    const filtered = groupAndFilterConversations(mockSearchHistories, 'ADVICE');
+    expect(filtered['Today'][0].title).toBe('Bitcoin Discussion');
+  });
+
+  it('should handle partial matches', () => {
+    const filtered = groupAndFilterConversations(
+      mockSearchHistories,
+      '  block',
+    );
+    expect(filtered['Last week'][0].title).toBe('Blockchain Overview');
+  });
+
+  it('should return all conversations when search term is empty', () => {
+    const filtered = groupAndFilterConversations(mockSearchHistories, '');
+
+    expect(filtered['Today'][0].title).toBe('Bitcoin Discussion');
+    expect(filtered['Yesterday'][0].title).toBe('Ethereum Chat');
+    expect(filtered['Last week'][0].title).toBe('Blockchain Overview');
+    expect(filtered['Last week'][1].title).toBe('Party planning tips');
```

Comment:
We get back an object with the time groups as keys, and the values are arrays of lighweight searchable conversation objects. This mirrors what have in the ChatHistory component before mapping out the JSX.